### PR TITLE
test: increase coverage for pkg/workers

### PR DESCRIPTION
**Pull request by JasonPowr**
## Description
Increase coverage for pkg/workers
jira: https://issues.redhat.com/browse/MGDSTRM-7990

## Verification Steps
1. Run `make test`
2. Run `make test/html/coverage/report `
3. Should be able to see the increased code coverage

## Checklist (Definition of Done)

- [x] All acceptance criteria specified in JIRA have been completed
- [x] Unit and integration tests added that prove the fix is effective or the feature works (tested against emulated and non-emulated OCM environment)
- [ ] ~~Documentation added for the feature~~
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer
- [ ] All PR comments are resolved either by addressing them or creating follow up tasks
- [ ] ~~Required metrics/dashboards/alerts have been added (or PR created).~~
- [ ] ~~Required Standard Operating Procedure (SOP) is added.~~
- [ ] ~~JIRA has been created for changes required on the client side~~

### CODE REVIEW

File: pkg/workers/leader_election_mgr_test.go
Comment (machi1990) [inline pkg/workers/leader_election_mgr_test.go:299]: There is no assertion here, is this desired? If so, why?

File: pkg/workers/leader_election_mgr_test.go
Comment (machi1990) [inline pkg/workers/leader_election_mgr_test.go:406]: There is no assertion here, why?

File: pkg/workers/leader_election_mgr_test.go
Comment (machi1990) [inline pkg/workers/leader_election_mgr_test.go:548]: let's use `gomega` for assertion

File: pkg/workers/reconciler_config_test.go
Comment (machi1990) [inline pkg/workers/reconciler_config_test.go:36]:
```suggestion
			name: "should always return nil ",
```

File: pkg/workers/worker_interface_test.go
Comment (machi1990) [inline pkg/workers/worker_interface_test.go:149]:
```suggestion
		{
			name: "should return true if the worker is running",
			fields: fields{
				isRunning: true,
			},
			want: true,
		},
		{
			name: "should return false if the worker is not running",
			fields: fields{
				isRunning: false,
			},
			want: false,
		},
```

File: pkg/workers/worker_interface_test.go
Comment (machi1990) [inline pkg/workers/worker_interface_test.go:221]: Why start and then stop the worker? Shouldn't this be testing start and stop methods separately?

File: pkg/workers/worker_interface_test.go
Comment (machi1990) [inline pkg/workers/worker_interface_test.go:266]: No any assertions are performed here, is this desired?

File: pkg/workers/leader_election_mgr_test.go
Comment (JasonPowr) [reply pkg/workers/leader_election_mgr_test.go:299]: Mainly because the method does not return anything to check, So I wasnt sure maybe I could check wantErr?

File: pkg/workers/leader_election_mgr_test.go
Comment (ziccardi) [inline pkg/workers/leader_election_mgr_test.go:223]: NewWithT is meant to be called for each subtest. Move this into `t.Run` please

File: pkg/workers/worker_interface_test.go
Comment (machi1990) [reply pkg/workers/worker_interface_test.go:221]: +1, let's separate them.

File: pkg/workers/reconciler_config_test.go
Comment (ziccardi) [inline pkg/workers/reconciler_config_test.go:20]:
```suggestion
	g.Expect(err).NotTo(HaveOccurred())
```

File: pkg/workers/reconciler_config_test.go
Comment (ziccardi) [inline pkg/workers/reconciler_config_test.go:45]: Move this into `t.Run(...`

File: pkg/workers/leader_election_mgr_test.go
Comment (ziccardi) [reply pkg/workers/leader_election_mgr_test.go:299]: Moreover, you could pass some `mock worker` instead of an empty array and check that the worker method are invoked with the correct params

File: pkg/workers/leader_election_mgr_test.go
Comment (ziccardi) [inline pkg/workers/leader_election_mgr_test.go:299]: Here you are just calling the mock function you just created. I don't think this is what you wanted to do.
I think would be better to add something like: 
```suggestion
			g.Expect(tt.fields.workers[0].calls.GetID).To(HaveLen(1))
			g.Expect(tt.fields.workers[0].calls.GetWorkerType).To(HaveLen(1))
```

File: pkg/workers/leader_election_mgr_test.go
Comment (machi1990) [inline pkg/workers/leader_election_mgr_test.go:300]: we also need to check if `.Start()` was called.

File: pkg/workers/leader_election_mgr_test.go
Comment (machi1990) [inline pkg/workers/leader_election_mgr_test.go:356]: let's add some assertion after this call as well

e.g check if `.Stop()` was called etc

File: pkg/workers/leader_election_mgr_test.go
Comment (JasonPowr) [reply pkg/workers/leader_election_mgr_test.go:247]: I attempted to add more test cases for the tests in this file  but every time I tried I kept running into issues with the sync.WaitGroup{}. It would constantly panic with the error `panic: sync: negative WaitGroup counter`. I was able to get around this by using .add(). But doing it this way broke the linter eg `copylocks: literal copies lock value from *tt.fields.workerGrp: sync.WaitGroup contains sync.noCopy`

File: pkg/workers/leader_election_mgr_test.go
Comment (pawelpaszki) [reply pkg/workers/leader_election_mgr_test.go:356]: Even though we will only confirm what we set in the mocks, `IsRunning()` can be checked like this:

```
	for _, testcase := range tests {
		tt := testcase
		t.Run(tt.name, func(t *testing.T) {
			g := NewWithT(t)
			t.Parallel()
			s := &LeaderElectionManager{
				workers:   tt.fields.workers,
				tearDown:  tt.fields.tearDown,
				workerGrp: sync.WaitGroup{},
			}
			if (len(s.workers)) > 0 {
				g.Expect(s.workers[0].IsRunning()).To(BeTrue())
			}
			s.Stop()
		})
	}
```

I find the workers stuff tricky to test and I am not familiar with this part of kfm code

Comment (ziccardi) [reply pkg/workers/leader_election_mgr_test.go:356]: @pawelpaszki Not sure `g.Expect(s.workers[0].IsRunning()).To(BeTrue())`: that would simply invoke the mock `IsRunningFunc` method that we defined here: https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1172/files/13feb35a18ff857e20f695a7eaafe25f1aed2cc2#diff-f95dcc5a7c0365d81ab0b0499b14bb8222240cde002620e604fe4e02cb5f9d57R335

I believe the test here should check that the mock methods are correctly invoked. Maybe something like:
```go
g.Expect(s.workers[0].calls.Stop).To(HaveLen(1))
```

Comment (pawelpaszki) [reply pkg/workers/leader_election_mgr_test.go:356]: makes sense @ziccardi

File: pkg/workers/leader_election_mgr_test.go
Comment (JameelB) [inline pkg/workers/leader_election_mgr_test.go:11]: After the tech discussion we decided not to use `.` import alias for gomega any longer. Let's change each assert methods to use `gomega.` instead.

**Comment by machi1990** [comment]
ping @ziccardi @pawelpaszki can you have another look at this PR?

**Comment by machi1990** [comment]
lgtm, thanks @JasonPowr 

let's wait for a review from @ziccardi and @pawelpaszki before merging

File: pkg/workers/leader_election_mgr_test.go
Comment (ziccardi) [inline pkg/workers/leader_election_mgr_test.go:508]: If `tearDown` is nil, no tests are performed, making the first test ("should return if teardown is nil") meaningless

File: pkg/workers/worker_interface_test.go
Comment (ziccardi) [inline pkg/workers/worker_interface_test.go:39]: Most of the tests here show that a getter returns the value set a couple of lines above.
I'm not sure it makes sense to test getter/setters.
IMO, the only useful tests here are the tests for `StartWorker` and `StopWorker`

@JasonPowr @machi1990 @pawelpaszki WDYT?

Comment (pawelpaszki) [reply pkg/workers/worker_interface_test.go:39]: I am on two minds here. getters/ setters should probably be tested if they are complex enough (rather then just setting/ getting values from some constants). I don't have a strong preference TBH. since the tests area already written I guess there is no harm in keeping them. WDYT?

**Review by ziccardi — COMMENTED** [review]
LGTM. Great job!

**Review by ziccardi — APPROVED** [review]

**Review by pawelpaszki — APPROVED** [review]

**Pull request merged**
